.circleci/config.yml

```diff
@@ -584,6 +589,49 @@ commands:
             cd << parameters.repo_path >>/study-builder
             mvn test-compile --batch-mode -DskipTests -Dmaven.repo.local=<< parameters.m2_repo >>
 
+  restore-ddp-cache:
+    description: Restore ddp cache
+    steps:
+      - run:
+          name: Download ddp-cache-md5 artifact
+          command: |
+            ddp_cache_md5='/home/circleci/.ddp-cache-md5.txt'
+            latest_artifact_url="https://circleci.com/api/v1.1/project/github/broadinstitute/ddp-study-server/latest/artifacts?circle-token=$CIRCLE_TOKEN"
+            md5_artifact_url=$(curl -sL "$latest_artifact_url" | jq -r '.[] | select(.node_index == 0) | select(.path == "ddp-cache-md5.txt") | .url')
+            if [[ -n "$md5_artifact_url" ]]; then
+              echo "Downloading artifact file [$md5_artifact_url] to file [$ddp_cache_md5]"
+              curl -sL "$md5_artifact_url?circle-token=$CIRCLE_TOKEN" -o "$ddp_cache_md5"
+            else
+              echo "No artifact file found, creating empty file [$ddp_cache_md5]"
+              touch "$ddp_cache_md5"
+            fi
+      - restore_cache:
+          keys:
+            - ddp-cache-{{ checksum "/home/circleci/.ddp-cache-md5.txt" }}
+            - ddp-cache-
+
+  save-ddp-cache:
+    description: Save ddp cache
+    steps:
+      - run:
+          name: Compute ddp-cache-md5
+          command: |
+            ddp_cache_dir='/home/circleci/.ddp-cache'
+            ddp_cache_md5='/home/circleci/.ddp-cache-md5.txt'
+            echo -n "" > "$ddp_cache_md5"
+            for cached_file in $(ls "$ddp_cache_dir" | sort); do
+              md5sum "$ddp_cache_dir/$cached_file" >> "$ddp_cache_md5"
+            done
+            echo "Contents of [$ddp_cache_md5]:"
+            cat "$ddp_cache_md5"
+      - save_cache:
+          key: ddp-cache-{{ checksum "/home/circleci/.ddp-cache-md5.txt" }}
+          paths:
+            - /home/circleci/.ddp-cache
+      - store_artifacts:
+          path: /home/circleci/.ddp-cache-md5.txt
+          destination: ddp-cache-md5.txt
```

Comment:
Not really understanding how storing the m5s is helping us. Would storing and then restoring the cache in whatever the last state it was be sufficient? Maybe just use the TestUserUtil.java file to generate a checksum like we do for pom.xml file to save the Java libraries? And if we have cache misses and have to get create users, could we just make sure that we enforce we handle the rate limiting (e.g., we get the rate limiting error, we try again a couple of seconds later).

Reply:
CircleCI's caching feature is great for caching dependencies. The `pom.xml` file specifies exactly what the cache contents are, so doing `{{ checksum "pom.xml" }}` gives us the cache behavior we want. But it's a different story when you want to cache some arbitrary stuff. So what I ended up doing was computing the md5 of the cache content into a file so we can use that file in the `checksum`.